crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,3298 @@
+                                BinaryOp::Mul => {
+                                    Some(_mul(SymbolExpr::Value(self * v), l.as_ref().clone()))
+                                }
+                                BinaryOp::Div => {
+                                    Some(_mul(SymbolExpr::Value(self / v), l.as_ref().clone()))
+                                }
+                                _ => None,
+                            },
+                            _ => None,
+                        },
+                    }
+                }
+            }
+            _ => None,
+        }
+    }
+
+    fn div_opt(&self, rhs: &SymbolExpr, recursive: bool) -> Option<SymbolExpr> {
+        match rhs {
+            SymbolExpr::Value(r) => Some(SymbolExpr::Value(self / r)),
+            SymbolExpr::Unary {
+                op: UnaryOp::Neg,
+                expr,
+            } => {
+                if recursive {
+                    self.div_opt(expr, recursive).map(_neg)
+                } else {
+                    None
+                }
+            }
+            SymbolExpr::Binary { op, lhs: l, rhs: r } => match l.as_ref() {
+                SymbolExpr::Value(v) => match op {
+                    BinaryOp::Mul => Some(_div(SymbolExpr::Value(self / v), r.as_ref().clone())),
+                    BinaryOp::Div => Some(_mul(SymbolExpr::Value(self / v), r.as_ref().clone())),
+                    _ => None,
+                },
+                _ => match r.as_ref() {
+                    SymbolExpr::Value(v) => match op {
+                        BinaryOp::Mul => {
+                            Some(_div(SymbolExpr::Value(self / v), l.as_ref().clone()))
+                        }
+                        BinaryOp::Div => {
+                            Some(_div(SymbolExpr::Value(self * v), l.as_ref().clone()))
+                        }
+                        _ => None,
+                    },
+                    _ => None,
+                },
+            },
+            _ => None,
+        }
+    }
+
+    pub fn opt_complex(&self) -> Option<Value> {
+        match self {
+            Value::Complex(c) => {
+                if (-SYMEXPR_EPSILON..SYMEXPR_EPSILON).contains(&c.im) {
+                    Some(Value::Real(c.re))
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    }
+}
+
+impl From<f64> for Value {
+    fn from(v: f64) -> Self {
+        Value::Real(v)
+    }
+}
+
+impl From<i64> for Value {
+    fn from(v: i64) -> Self {
+        Value::Int(v)
+    }
+}
+
+impl From<Complex64> for Value {
+    fn from(v: Complex64) -> Self {
+        if (-SYMEXPR_EPSILON..SYMEXPR_EPSILON).contains(&v.im) {
+            Value::Real(v.re)
+        } else {
+            Value::Complex(v)
+        }
+    }
+}
+
+impl Add for Value {
+    type Output = Value;
+    fn add(self, rhs: Self) -> Value {
+        &self + &rhs
+    }
+}
+
+impl Add for &Value {
```

Comment:
`Value` I think actually _should_ be `Copy`. Perhaps you can add `#[derive(Copy)]` for `Value` and move the logic below into the `Add` impl for `Value`.

Same goes for the other math op impls here.